### ISSUE TITLE
Thoughts for a new API

### ISSUE THREAD

**Issue by endel** [endel/delta-listener#5]
I feel the usage of this library should be more intuitive than it is right now.

### Current API:

```typescript
public listen(path: string, operation: string, callback: Function);
```
The problems I see with it:

- The `operation` parameter is usually not essential, and it causes more confusion than helps. The callback for `"add"` and `"replace"` operations are often the same anyway (from my experience)
- The number of arguments in the `callback` varies according to the number of matches the `path` has - which can be confusing. (Which argument number is the `value`? BTW `"removed"` operations doesn't have a `value`! 😖 )

In order to address these issues, I'd like to propose some possible solutions.

###  Omitting  the `operation` parameter

By omitting the `operation` parameter, the callback should receive its value somehow. Given the current behaviour of the callback, it could be used as the last argument.

```typescript
container.listen(
  "entities/:id/:attribute",
  (entityId: string, attribute: string, value: any, operation: string) => {
    console.log(entityId, attribute, operation, value);
  }
);
```

### Having a standard for the arguments in the `callback`

By having the same signature in the callback for any kind of change (`"add"`, `"remove"`, `"replace"`), all the confusion should be gone, and it plays nicer with TypeScript.

```typescript
interface DataChange {
    path: string[];
    value: any;
    operation: string;
}

container.listen("entities/:id/:attribute", (change: DataChange) => {
    // Destructing the variables from the path (this pattern may repeat a lot)
    let [ entityId, attribute ] = change.path;

    console.log(entityId, attribute, change.operation, change.value);
})
```

A variation of this one would be to have the `path` as an object, already populated with the named segments. 

```typescript
interface DataChange {
    path: any;
    value: any;
    operation: string;
}

container.listen("entities/:id/:attribute", (change: DataChange) => {
    let entityId = change.path.id; // extracted from `:id` in the first argument
    let attribute = change.path.attribute; // extracted from `:attribute` in the first argument
    console.log(entityId, attribute, change.operation, change.value);
})
```

### Requesting feedback 😇 

I'd love to gather some feedback before changing the API, since this library has been ported to [C#](https://github.com/gamestdio/colyseus-unity3d) and [C++](https://github.com/chunho32/colyseus-cocos2d-x). Would be awesome to have a solution that everybody's happy about :)

Would you recommend a different approach? /cc @AnubisCode @seiyria @niklaswolf @darkyen @michalhainc @bmovement @JoaoMosmann @chunho32

Sorry for the long post, thanks for reading all this :)

**Comment by michalhainc** [comment]
I think these proposals are build on logical arguments, at this moment I have no objections/improvement ideas to this matter

**Comment by seiyria** [comment]
I like the idea of having a path and no operation. For me, some code looks like this:

```
    this.worldRoom.state.listen('players/:id/effects/:effect', 'add', (entityId, effect, value) => {
      this.clientGameState.updatePlayerEffect(+entityId, effect, value);
    });

    this.worldRoom.state.listen('players/:id/effects/:effect', 'remove', (entityId, effect) => {
      this.clientGameState.updatePlayerEffect(+entityId, effect, null);
    });
```

Simplifying it would be great.

There are a few things like that. I know I ran into a confusing point, as well, with this code: 
```
    this.worldRoom.state.listen('mapData/openDoors/:id/isOpen', 'add', updateDoor);
    this.worldRoom.state.listen('mapData/openDoors/:id/isOpen', 'replace', updateDoor);
```

The add never fired when adding a new door with the isOpen property (`openDoors` is an object), instead, to make it work, I had to do this:

```
    this.worldRoom.state.listen('mapData/openDoors/:id', 'add', updateDoor);
    this.worldRoom.state.listen('mapData/openDoors/:id/isOpen', 'replace', updateDoor);
```

Also, I don't know how feasible it is to request and send batched updates, but I have code like this:
```
    this.worldRoom.state.listen('players/:id/:attribute', 'replace', (entityId, attribute, value) => {
      if(!includes(['x', 'y', 'dir', 'swimLevel'], attribute)) return;
      this.clientGameState.updatePlayer(+entityId, attribute, value);
    });
```

Every update will fire at least twice, because of x and y. x and y may not change (moving along an axis), but most often I get two updates where I could get one. I don't know what an API for this could look like, though. It's not the end of the world, but if it was something that could be factored in, that would be great.

**Comment by niklaswolf** [comment]
I really like the last variation, where `path` is an already populated object, simply because it requires less work for the developer. In every case you would have to destruct the variables from path anyway, so I like the approach to do this automatically. It looks more cleaner and understandable too I think.
Furthermore the general approach of only having one parameter for the callback makes it much easier having very "long" pathes. In the current implementation you'd have to provide a parameter for every path-part. In the new approach this would be much easier as it would be done automatically!

**Comment by AnubisCode** [comment]
about the last variation:
looks like not the most productive. to checking `includes(['x', 'y', 'dir', 'swimLevel'], attribute)` every updates - veeeeery no effective.

**Comment by seiyria** [comment]
It's not the best idea, I did it to get around this issue: https://github.com/endel/delta-listener/issues/4 - I could change it back. But it's still a case where I have to watch x, y, dir, swimlevel, and bind the change to them 4 different times.

**Comment by seiyria** [comment]
Another thing that I noticed is that variables prefixed with a dollar sign do not get sent across. I don't know if that's intentional, but it's not documented anywhere.

I use $var to note that "this is a temporary variable and should not be saved, but is relevant in some way." I can always use a different prefix of course but if that's intentional (or me doing something wrong) that should at least be documented. :D

**Comment by endel** [comment]
Thanks for the feedback guys! I really appreciate your time to reply here :)

This new approach is actually 1.3x faster than the previous one. When the new version of Colyseus comes out this new API will be there as well :) Will try to provide some docs for upgrading.

@seiyria I like the way you're approaching the listeners. I've been thinking lately that it would be awesome to build a another tool on top of `delta-listener` to allow automatic property changes, by using decorators - somewhat similar to how [Unity3D's state synchronization](https://docs.unity3d.com/Manual/UNetStateSync.html) works:

```typescript
import { sync } from "sync-tool"

class Entity {
    @sync x: number;
    @sync y: number;
}
```

The challenge here is how to infer to object scope based on the room's data. I'm totally interested in implementing this eventually.

**Comment by seiyria** [comment]
@endel you have no idea how awesome something like `@sync` would be! It's basically what I would want - the sugar looks nice and it would be SO much nicer on the client side.

**Comment by darkyen** [comment]
@endel why do that and not use Observables instead?

**Comment by endel** [comment]
@darkyen I don't have much experience using Observables. I see that a uncompressed implementation has 11kb (https://github.com/zenparsing/zen-observable/blob/master/src/Observable.js)

If the benefits of using it pay off, why not? :D 

Do you have some example how the API usage may look like using Observables?

**Comment by JoaoMosmann** [comment]
Well, I don't have much to argue with. But the changes being a single `Object`, and also containing the path as an `Object` is great, much less prone to mistakes than the `Array` and parameters variations.

**Comment by seiyria** [comment]
One thing that I definitely want to make note of, since I'm running into it right now. If I do something like this:

this.worldRoom.state.listen('players/:id/agro', 'replace', updateSpecificAttr.bind(this, 'agro')); and `agro` is an object, I'm hoping that `replace` (or the new equivalent) would also watch the keys inside of it, somehow, and send me updates for any change for this object. Otherwise, I have to:

```
    this.worldRoom.state.listen('players/:id/agro/:player', 'add', updateSpecificAttr.bind(this, 'agro'));
    this.worldRoom.state.listen('players/:id/agro/:player', 'replace', updateSpecificAttr.bind(this, 'agro'));
    this.worldRoom.state.listen('players/:id/agro/:player', 'remove', updateSpecificAttr.bind(this, 'agro'));
```

And I don't care about `/:player`, it's just there to grab all the updates from the object.

**Comment by endel** [comment]
Thanks guys! 

- Released [`delta-listener@2.0.0`](https://github.com/endel/delta-listener/releases/tag/2.0.0) with this change. 
- Released [`colyseus.js@0.7.0-alpha.4`](https://github.com/gamestdio/colyseus.js/releases/tag/0.7.0-alpha.4) using it. 

(Next steps will be updating the C# and C++ implementations.)

Maybe for `3.0.0` we can use Observables. Making slow and incremental changes :)
